### PR TITLE
CUDA: add dynamic shared mem to softmax, refactor general usage

### PR DESCRIPTION
**Pull request by am17an**
This PR does two things: 

1) Add the dynamic shared memory to softmax. Added a perf case where shared_mem requried >= 48 kb but less than smpbo on my RTX 3090 (which is 102.4 kb)

PR
```
  SOFT_MAX(type=f32,ne=[12888,1024,5,1],mask=0,m_prec=f32,scale=1.000000,max_bias=0.000000):                    1320 runs -   765.44 us/run -   515520 kB/run -  647.16 GB/s
``` 
vs master
```
  SOFT_MAX(type=f32,ne=[12888,1024,5,1],mask=0,m_prec=f32,scale=1.000000,max_bias=0.000000):                     924 runs -  1126.70 us/run -   515520 kB/run -  439.66 GB/s
```

2) It also refactors the set dynamic memory routine to a macro for readability

### CODE REVIEW

File: ggml/src/ggml-cuda/common.cuh [hunk 1]
```diff
@@ -175,6 +175,20 @@ static const char * cu_get_error_str(CUresult err) {
...
+        static bool shared_memory_limit_raised[GGML_CUDA_MAX_DEVICES] = {false}; \
+        const int id = ggml_cuda_get_device(); \
+        if (!shared_memory_limit_raised[id]) { \
+            CUDA_CHECK(cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, nbytes)); \
+            shared_memory_limit_raised[id] = true; \
+        } \
+    } while (0)
+#else
+#define CUDA_SET_SHARED_MEMORY_LIMIT(kernel, nbytes) do {} while (0)
+#endif
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/common.cuh]:
```suggestion
#endif // !(defined(GGML_USE_HIP) && defined(__HIP_PLATFORM_AMD__)) && !defined(GGML_USE_MUSA)
```

File: ggml/src/ggml-cuda/cross-entropy-loss.cu [hunk 2]
```diff
@@ -123,13 +123,7 @@ void ggml_cuda_cross_entropy_loss(ggml_backend_cuda_context & ctx, ggml_tensor *
...
 
     if (nbytes_shared <= smpbo) {
-#if !(defined(GGML_USE_HIP) && defined(__HIP_PLATFORM_AMD__)) && !defined(GGML_USE_MUSA)
-        static bool shared_memory_limit_raised[GGML_CUDA_MAX_DEVICES] = {false};
-        if (!shared_memory_limit_raised[id]) {
-            CUDA_CHECK(cudaFuncSetAttribute(cross_entropy_loss_f32<true>, cudaFuncAttributeMaxDynamicSharedMemorySize, smpbo));
-            shared_memory_limit_raised[id] = true;
-        }
-#endif // !(defined(GGML_USE_HIP) && defined(__HIP_PLATFORM_AMD__)) && !defined(GGML_USE_MUSA)
+        CUDA_SET_SHARED_MEMORY_LIMIT((cross_entropy_loss_f32<true>), nbytes_shared);
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/cross-entropy-loss.cu]: I just noticed that (already on master) this logic is incorrect. For MMQ and FlashAttention the amount of shared memory used is constant, for the cross entropy loss it varies depending on `ne00` so it may need to be raised multiple times. It should be fine to just set the amount of usable shared memory to the maximum value instead.

File: ggml/src/ggml-cuda/softmax.cu [hunk 3]
```diff
@@ -150,6 +151,24 @@ static __global__ void soft_max_back_f32(
...
+
+        CUDA_SET_SHARED_MEMORY_LIMIT(
+            (soft_max_f32<true, ncols, block, half >),  smpbo);
+        CUDA_SET_SHARED_MEMORY_LIMIT(
+            (soft_max_f32<true, ncols, block, float>), smpbo);
+    };
+
+    //unary fold
+    ( apply_limit(std::integral_constant<int, Ns>{}), ... );
+}
```
Comment (JohannesGaessler) [inline ggml/src/ggml-cuda/softmax.cu]: Define a macro for a switch case instead? I would (slightly) prefer it to avoid duplicating the logic for setting the block size as a function of `ncols`. I think this would ultimately also be fine though.

**Review by JohannesGaessler — APPROVED** [review]

**Comment by JohannesGaessler** [comment]
(To be clear, the approval is conditional on not breaking the CI.)

**Comment by CISC** [comment]
> (To be clear, the approval is conditional on not breaking the CI.)

Unfair, it just broke itself (I'll restart). :)

**Comment by CISC** [comment]
The failing CIs can be safely ignored, it's just a DNS issue.
